### ISSUE TITLE
Some Questions about SSL Pre-training

### ISSUE THREAD

**Issue by ahkarami** [NVIDIA/NeMo#3792]
Hi, 
The SSL pre-training code & tutorial are very great; however, I have some questions about configurations of models in SSL pre-training (there are some ambiguities for me to set parameters correctly). I have prepared a `citrinet_ssl_512.yaml` config file for SSL pre-training of Citrinet512 model as below: (I have implement it based on the [Nemo citrinet_ssl_1024.yaml](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/ssl/citrinet/citrinet_ssl_1024.yaml) file. But, as Citrinet1024 is a little huge model I decided to switch to Citrinet512 model for my research experiments). 

[citrinet_ssl_512.zip](https://github.com/NVIDIA/NeMo/files/8178032/citrinet_ssl_512.zip)

My questions are as follows:
1- In the above uploaded file, I have written `doubt` comment beside some of parameters. Can you please kindly check these parameters and let me know if they are set correctly or not? I also suggest that add this yaml file (and also yaml config files for other ASR models in the [SSL conf folder](https://github.com/NVIDIA/NeMo/tree/main/examples/asr/conf/ssl) of NeMo (_e.g._, yaml config file for QuartzNet, Citrinet256, contextnet256, etc.)). 
2- If we set `precision: 16` is it adequate for training the model with fp16? or we need apply extra changes?
3- I have set `min_duration: 2.0`. But, in my data set there are audio utterances with duration less than 2 seconds. So, it means that all of these utterances will be removed form the training process? Is there any way to handle this issue and keep them? (I have realized that there is a relation between `spec_augment` parameters and  `min_duration`. As I have set `patch_size: 16` & `mask_patches: 10`, so 16*10=160, and it means that the minimum acceptable duration of an utterance must be 1.6 seconds. As a result, I have set `min_duration: 2.0`. Am I correct?)
4- I have set the `max_duration: 60.0`. Is this parameter important and related to `spec_augment` category parameters?
5- I have set `trim_silence: false` & `pin_memory: true` & add `num_workers: 16`. Do the values of these parameters have a special effect on the training process? 
6- I have set `num_negatives: 40` & `sample_from_same_utterance_only: true` & `sample_from_non_masked: false`. Do you think the values of these parameters are set correctly? I have set these values based on your [tutorial file](https://nbviewer.org/github/NVIDIA/NeMo/blob/main/tutorials/asr/Self_Supervised_Pre_Training.ipynb), and other `spec_augment` category parameters' values.
7- About loading `init_checkpoint` (_i.e._, `+init_from_pretrained_model`) I have previously send an [issue](https://github.com/NVIDIA/NeMo/issues/3728). What I learned from that is that to do this, I must first download the model manually from NGC and then enter the .nemo file address of the model to code. Did I get it right?
In fact, I want to evaluate the impact of loading another model's weight at the beginning of SSL process for my self, and I know that it is not a formal acceptable task.
8- After the process of SSL training has been finished, can one extract the log probability of the SSL model for evaluation? I mean that, can we extract log probability of model for each word (or token) to for example evaluate (analyze) which words the model transcribe with more skepticism and which words it is more confident about transcribing. Can you help me with this too?

I apologize to you that my number of questions has increased and I thank you in advance if you can guide me.
Best

**Comment by sam1373** [comment]
1, 3, 4, 6. If the durations in your data vary by more than a few seconds, I would recommend to re-cut the audio into more uniform-lengthed segments for the self-supervised training, and then set the mask_patches to cover around 50% of the new segment duration (for example, cut to 15 seconds and set mask_patches to 15x48, which would mean a total of ~7 masked seconds). Otherwise, if you set min_duration to 2 and max_duration to 60, you will have to set mask_patches to low enough to work for 2 seconds, however that would mean that when processing longer segments you would still only cut out a small amount of patches from them, which would be wasteful. You would also waste a lot of processing power on shorter segments since you would be adding a lot of padding to match the longer ones in the same batch. If your samples are too short you also run into the problem of having to sample less negatives, which degrades the results.
2. Yes, you can use fp16, although note that in some cases training with it may be unstable.
5. These parameters look good.
7. Yes, you can load weights from the downloaded .nemo file.
8. The model after ssl pre-training without any additional supervised fine-tuning will not be able to transcribe audio, or produce log-probabilities for tokens, so you can't evaluate it that way.

**Comment by ahkarami** [comment]
@sam1373 
Thanks for your great and complete answers. 
1- about the question of setting `min_duration`, what I have realized from your answer is that, for example split audio samples (utterances) to e.g., 15 seconds uniform-lengthed samples and set `mask_patches` to 15x48. 
1-1: Is this 15 seconds length an appropriate option as your opinion (or is it better to increase or decrease its amount? and, is this related to the processing power and memory of our GPUs?) 
1-2: It is clear that these 15 seconds length audio samples are not segmented correctly (I mean, the audio file may have been cut while the person was talking). Doesn't this pose a serious problem for the SSL training?
1-3: for example, if we cut all audio samples as exactly 15 seconds utterances, and set `mask_patches` to 15x48; then is it appropriate to set `min_duration: 15.0` & `max_duration: 15.0` & `stride_layers: 1` & `combine_time_steps: 4` & `num_negatives: 180`? I mean how we can specify the `num_negatives` value? I set it to 180, because 15x8=4x180. Am I correct?
1-4: Can you please explain for me that what exactly the `min_duration` & `max_duration` parameters do? In the training process the code will just consider utterances with length between `min_duration` & `max_duration` and remove other ones?
------- ---------- ----------
2- About the question of `fp16`, can you please explain what do you mean about `unstable training in some cases`? Do you mean that the value of val_loss (or train_loss) changes oscillating? And does not decrease normally and regularly? To address this problem if we set `warmup_steps: 5000` & monitor the model's behavior while training via TensorBoard is it adequate? 
----
thanks you in advance for your help.
Best

**Comment by sam1373** [comment]
1. If a large part of your data consists of samples that are too short to cut properly into 15 second segments, you can pick a lower duration (adjust mask_patches accordingly). Min_duration and max_duration parameters will not do anything apart from excluding samples with duration outside of the range. By default num_negatives is set to 100, which is a decent amount; you can try a higher amount if you have enough masked steps.
2. I meant that fp16 training is more likely to diverge
